src/model_repository_manager.cc

```diff
@@ -1803,27 +1818,57 @@ ModelRepositoryManager::Poll(
     // [DLIS-3487] Model doesn't need to be in repository if model files
     // are provided in flight.
     for (const auto& model : models) {
+      // Check model mapping first to see if matching model to load.
       bool exists = false;
-      for (const auto repository_path : repository_paths_) {
-        bool exists_in_this_repo = false;
-        const auto full_path = JoinPath({repository_path, model.first});
-        Status status = FileExists(full_path, &exists_in_this_repo);
-        if (!status.IsOk()) {
-          LOG_ERROR << "failed to poll model repository '" << repository_path
-                    << "' for model '" << model.first
-                    << "': " << status.Message();
+      auto model_it = model_mappings_.find(model.first);
+      if (model_it != model_mappings_.end()) {
+        auto res =
+            model_to_repository.emplace(model.first, model_it->second.first);
+        if (res.second) {
+          exists = true;
+        } else {
+          exists = false;
+          model_to_repository.erase(res.first);
+          LOG_ERROR << "failed to poll model '" << model.first
+                    << "': not unique across all model repositories";
           *all_models_polled = false;
-        } else if (exists_in_this_repo) {
-          auto res = model_to_repository.emplace(model.first, repository_path);
-          if (res.second) {
-            exists = true;
-          } else {
-            exists = false;
-            model_to_repository.erase(res.first);
-            LOG_ERROR << "failed to poll model '" << model.first
-                      << "': not unique across all model repositories";
+        }
+      } else {
+        for (const auto repository_path : repository_paths_) {
+          bool exists_in_this_repo = false;
+          const auto full_path = JoinPath({repository_path, model.first});
+          Status status = FileExists(full_path, &exists_in_this_repo);
+          if (!status.IsOk()) {
+            LOG_ERROR << "failed to poll model repository '" << repository_path
+                      << "' for model '" << model.first
+                      << "': " << status.Message();
             *all_models_polled = false;
-            break;
+          } else if (exists_in_this_repo) {
+            // Check to make sure this directory is not mapped.
+            // If mapped, continue to next repository path.
+            bool mapped = false;
+            for (auto const& mapping : model_mappings_) {
+              if (mapping.second.second == full_path) {
+                mapped = true;
+                break;
+              }
+            }
+            if (mapped) {
+              continue;
+            }
```

Comment:
Shouldn't need this, if control reaches this point there shouldn't be an existing mapping. `else` in line 1836 is equivalent to `model_mappings_.find(model.first) == model_mappings_.end()`

Reply:
This is an edge case, revealed by one of the tests (DifferentMappings, I believe). Say we have two repositories: models_0 with model_0 mapped to "name_0," and models_1 with model_0 and model_1 (no mappings). If you don't add this check, model_0 from models_0 could be loaded, yet that's counter to the behavior the user requested by mapping that subdirectory to a different name ("name_0").

Reply:
I see, resolved